src/controllers/controllerengine.cpp

```diff
@@ -98,30 +98,48 @@ void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefix
     }
 }
 
-/* -------- ------------------------------------------------------
-Purpose: Resolves a function name to a QScriptValue including
-            OBJECT.Function calls
-Input:   -
-Output:  -
--------- ------------------------------------------------------ */
-QScriptValue ControllerEngine::resolveFunction(const QString& function) const {
-    QHash<QString, QScriptValue>::const_iterator i =
-            m_scriptValueCache.find(function);
-    if (i != m_scriptValueCache.end()) {
-        return i.value();
-    }
+/* ------------------------------------------------------------------
+Purpose: Turn a snippet of JS into a QScriptValue function.
+         Wrapping it in an anonymous function allows any JS that
+         evaluates to a function to be used in MIDI mapping XML files
+         and ensures the function is executed with the correct
+         'this' object.
+Input:   QString snippet of JS that evaluates to a function,
+         int number of arguments that the function takes
+Output:  QScriptValue of JS snippet wrapped in an anonymous function
+------------------------------------------------------------------- */
+QScriptValue ControllerEngine::wrapFunctionCode(const QString& codeSnippet,
+                                                int numberOfArgs) {
+    QScriptValue wrappedFunction;
 
-    QScriptValue object = m_pEngine->globalObject();
-    QStringList parts = function.split(".");
+    QHash<QString, QScriptValue>::const_iterator i =
+            m_scriptWrappedFunctionCache.find(codeSnippet);
 
-    for (int i = 0; i < parts.size(); i++) {
-        object = object.property(parts.at(i));
-        if (!object.isValid()) {
-            break;
+    if (i != m_scriptWrappedFunctionCache.end()) {
+        wrappedFunction = i.value();
+    } else {
+        QScriptValue function = m_pEngine->evaluate(codeSnippet);
```

Comment:
Evaluating codeSnippet can produce benign errors (as you pointed out -- for the case when the script author hasn't defined the handler yet) so I think we shouldn't be evaluating codeSnippet here because it means we will cache the failure permanently. 

Instead, here's what I'm proposing:
```c++
// Do not evaluate(codeSnippet).
if (!syntaxIsValid(codeSnippet)) {
  wrappedFunction = QScriptValue();
} else {
  // build wrappedCode (elided)
  wrappedFunction = m_pEngine->evaluate(wrappedCode);
  checkException();
}
m_scriptWrappedFunctionCache[codeSnippet] = wrappedFunction;
```

If an exception occurs, it's either because:
* codeSnippet has a syntax error
* codeSnippet is a statement, not an expression (wrappedFunction evaluation will fail in that case with a syntax error).

In the former case, the script dialog will pop from the syntaxIsValid check. In the latter case the script dialog will pop from the checkException. 

The return value will be a function (the wrapper) on success and an invalid QScriptValue if either of the syntax error cases occurred. If we return a non-function then it'll be dropped by internalExecute() and we've already notified the user via script dialog so it doesn't need double-handing.

In the event of a transient/benign failure, the execute() following wrapFunctionCode() in MidiController will show a script error dialog with the transient exception. If the script fixes itself, then the next button press will succeed.